### PR TITLE
Added iterator agent for ADK pipeline #1611

### PR DESCRIPTION
**Pull request by vivek100**
Add ArrayIteratorAgent for Array Iteration Functionality

Summary
Implements ArrayIteratorAgent to provide built-in array iteration functionality for the ADK. This agent allows developers to iterate over arrays in session state and pass each element to a sub-agent for processing, addressing a common workflow pattern that was previously missing from the ADK.

Changes Made

✅ Added ArrayIteratorAgent class in src/google/adk/agents/array_iterator_agent.py
✅ Implemented single sub-agent constraint with validation
✅ Added support for nested array key access using dot notation (e.g., user_data.customers)
✅ Implemented automatic result collection from sub-agent's output_key
✅ Added comprehensive unit tests in tests/unittests/agents/test_array_iterator_agent.py
✅ Updated agent exports and imports for proper integration
✅ Added documentation and examples
Key Features

Single Sub-Agent Pattern: Enforces exactly one sub-agent to keep iteration focused and simple
Nested State Access: Supports dot notation for accessing nested arrays (app:config.items, user.data.list)
Automatic Collection: Collects results from sub-agent's output_key into an array
Configurable Context: Passes each array element via configurable item_key (default: current_item)
Optional Limits: Supports max_items parameter for processing subsets
Robust Error Handling: Validates array existence, sub-agent count, and handles edge cases
Usage Example

# Sub-agent processes individual items
processor = LlmAgent(
    name="item_processor", 
    instruction="Process: {current_item}",
    output_key="result"
)

# Array iterator
iterator = ArrayIteratorAgent(
    name="batch_processor",
    array_key="items",
    output_key="all_results", 
    sub_agents=[processor]
)
Testing

✅ Unit tests cover all core functionality
✅ Edge cases tested (empty arrays, missing keys, validation errors)
✅ Integration with existing ADK patterns verified
✅ Nested state access thoroughly tested
✅ Error handling and validation confirmed
Breaking Changes
None - this is a new feature addition that doesn't modify existing functionality.

Documentation

Added comprehensive docstrings following Google style
Included usage examples in class documentation
Added type hints for all parameters and methods
Fixes https://github.com/google/adk-python/issues/1610

### CODE REVIEW

File: src/google/adk/agents/array_iterator_agent.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+      try:
+        # For simple keys, use direct assignment
+        if '.' not in self.output_key:
+          ctx.session.state[self.output_key] = results
+        else:
+          # For nested keys, we need to work with the state dict
+          state_dict = ctx.session.state.to_dict()
+          _set_nested_value(state_dict, self.output_key, results)
+          # Update the session state with the modified dict
+          ctx.session.state.update(state_dict)
```
[the remark this answers is not in the document]
Comment (vivek100) [reply src/google/adk/agents/array_iterator_agent.py]: Can you create a code suggestion with the fix please?

@gemini-code-assist

File: tests/unittests/agents/test_array_iterator_agent.py [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+            _set_nested_value(data, "user.profile.name", "John")
+    
+    def test_set_nested_value_empty_key(self):
+        data = {}
+        with pytest.raises(ValueError, match="Key path cannot be empty"):
+            _set_nested_value(data, "", "value")
+
+
+class TestArrayIteratorAgent:
+    """Test the ArrayIteratorAgent class."""
```
[the remark this answers is not in the document]
Comment (vivek100) [reply tests/unittests/agents/test_array_iterator_agent.py:161]: @gemini-code-assist can you create the test file to test this please?

[the remark this answers is not in the document]
Comment (vivek100) [reply tests/unittests/agents/test_array_iterator_agent.py:161]: can you create code suggestion or share the code for the test case please?

@gemini-code-assist

[the remark this answers is not in the document]
Comment (vivek100) [reply tests/unittests/agents/test_array_iterator_agent.py:161]: Added the test can you review the PR again please @gemini-code-assist

File: contributing/samples/array_iterator_agent/README.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+### 2. Nested Data Processing
+
+```python
+# Process nested customer array
+customer_processor = ArrayIteratorAgent(
+    name="customer_processor",
+    array_key="company.customers",        # Nested array access
+    item_key="current_customer",
+    output_key="company.processed",       # Nested output
+    sub_agents=[customer_analyzer]
```
[the remark this answers is not in the document]
Comment (vivek100) [reply contributing/samples/array_iterator_agent/README.md]: Can you share code suggestions for all the instances please? @gemini-code-assist

File: src/google/adk/agents/array_iterator_agent.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    
+    # Store results if output_key is specified
+    if self.output_key and results is not None:
+      try:
+        # For simple keys, use direct assignment
+        # For both simple and nested keys, we can work with the state dict
+        state_dict = ctx.session.state.to_dict()
+        _set_nested_value(state_dict, self.output_key, results)
+        # Update the session state with the modified dict
+        ctx.session.state.update(state_dict)
```
[the remark this answers is not in the document]
Comment (vivek100) [reply src/google/adk/agents/array_iterator_agent.py]: Thanks updated the comment

File: contributing/samples/array_iterator_agent/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (vivek100) [reply contributing/samples/array_iterator_agent/README.md]: Thanks added all the comment updates @gemini-code-assist

**Comment by vivek100** [comment]
@gemini-code-assist can you look at the current implementation of agents in the main branch and create a pros and cons list of whether the itterator agent i created is helpful or not or there are other features already in the agent which we can use to achieve this. The idea is to give this option out of the box and option to connect this iterator agent with any of the available agents.

**Comment by rohityan** [comment]
Hi @vivek100 ,Thank you for your contribution! 
We appreciate you taking the time to submit this pull request. Your PR has been received by the team and is currently under review. We will provide feedback as soon as we have an update to share.

**Comment by hangfei** [comment]
This fits better for the adk-python-community repo. please move it there. thanks. (https://github.com/google/adk-python-community)
